pkg/speaker/layer2/memberlist.go

```diff
@@ -0,0 +1,212 @@
+	config.Events = &memberlist.ChannelEventDelegate{Ch: eventCh}
+	list, err := memberlist.Create(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &layer2Speaker{
+		eventCh:    eventCh,
+		reloadChan: reloadChan,
+		mlist:      list,
+		client:     client,
+		queue:      queue,
+		announcers: map[string]announcer{}}, nil
+}
+
+func (l *layer2Speaker) joinMembers() error {
+	iplist := []string{}
+	pods, err := l.client.CoreV1().Pods(util.EnvNamespace()).List(context.TODO(), v1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(labels.Set{"app": "openelb", "component": "speaker"}).String(),
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, p := range pods.Items {
+		iplist = append(iplist, p.Status.PodIP)
+	}
+
+	_, err = l.mlist.Join(iplist)
+	return err
+}
+
+type layer2Speaker struct {
+	mlist      *memberlist.Memberlist
+	eventCh    chan memberlist.NodeEvent
+	reloadChan chan event.GenericEvent
+	client     *kubernetes.Clientset
+	queue      workqueue.Interface
+
+	// nic - announcers
+	announcers map[string]announcer
+}
+
+func (l *layer2Speaker) SetBalancer(ip string, clusterNodes []corev1.Node) error {
+	member := map[string]string{}
+	for _, m := range l.mlist.Members() {
+		member[m.Name] = m.Addr.String()
+	}
+
+	nodes := []string{}
+	for _, n := range clusterNodes {
+		if _, exist := member[n.GetName()]; exist {
+			nodes = append(nodes, n.GetName())
+		}
+	}
+
+	if len(nodes) == 0 {
+		klog.Warningf("no suitable nodes to participate in the announced election.")
+		return nil
+	}
+
+	// Sort the slice by the hash of node + load balancer ips. This
+	// produces an ordering of ready nodes that is unique to all the services
+	// with the same ip.
+	sort.Slice(nodes, func(i, j int) bool {
+		hi := sha256.Sum256([]byte(nodes[i] + "#" + ip))
+		hj := sha256.Sum256([]byte(nodes[j] + "#" + ip))
+
+		return bytes.Compare(hi[:], hj[:]) < 0
+	})
+
+	klog.Infof("node %s wins the right to announce the IP address %s", nodes[0], ip)
+	if nodes[0] != util.GetNodeName() {
+		return nil
+	}
+
+	for _, a := range l.announcers {
+		if a.ContainsIP(ip) {
+			return a.AddAnnouncedIP(ip)
+		}
+	}
+
+	return nil
+}
+
+func (l *layer2Speaker) DelBalancer(ip string) error {
+	for _, a := range l.announcers {
+		if a.ContainsIP(ip) {
+			return a.DelAnnouncedIP(ip)
+		}
+	}
+	return nil
+}
+
+func (l *layer2Speaker) Start(stopCh <-chan struct{}) error {
+	if err := l.joinMembers(); err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	go l.processEIP(ctx)
+	go func() {
+		tick := time.NewTicker(5 * time.Second)
+		for {
+			select {
+			case <-stopCh:
+				cancel()
+			case <-l.eventCh:
+				evt := corev1.Service{}
+				evt.Name = constant.Layer2ReloadServiceName
+				evt.Namespace = constant.Layer2ReloadServiceNamespace
+				l.reloadChan <- event.GenericEvent{Object: &evt}
+			case <-tick.C:
+				for _, member := range l.mlist.Members() {
+					klog.Infof("member info %s %s", member.Name, member.Address())
```

Comment:
I believe klog.Infof should add a log level, otherwise the logs here will output every 5 seconds. I feel there will be a lot of these repetitive information in the log file.

```suggestion
					klog.V(2).Infof("member info %s %s", member.Name, member.Address())
```

Reply:
done.